Removed CoPromise

belozierov opened the pull request:
- removed CoPromise
- updated await func

Please also compare the performance of this version versus the version in pull request. This version is lock-free but it suspends/resumes coroutine every time. I’m not good at NIO :) so please check this yourself.
```
    public func await() throws -> Value {
        let coroutine = try Coroutine.current()
        var awaitResult: Result<Value, Error>!
        coroutine.suspend {
            self.whenComplete { result in
                awaitResult = result
                coroutine.resume()
            }
        }
        return try awaitResult.get()
    }
```

calebkleveter commented:
I like how this is looking! There are a few superfluous things that I would change, but I can take care of that later.

I want to make sure I understand the purpose of this PR. Does it improve the performance of the `.await` method and remove potential leaks by ensuring that the coroutine will always pause and then resume again?

belozierov commented:
There are no leaks in all versions. I removed overhead usage of CoPromise, because EventLoopFuture is already used for the same purposes.

The version in PR uses NSLock for thread safety. In comments I added an alternative version of the await function. Please test which of them has better performance.

calebkleveter commented:
I have a couple of changes that I want to make before merging this into master. Thanks for getting this started!
